File: graph_algorithms/week4_paths_in_graphs2/1_minimum_flight_cost/dijkstra.py
from queue import PriorityQueue
from typing import List


def distance(adj: List[List[int]], cost: List[List[int]], s: int, t: int) -> int:  # noqa: C901
    dist: List[int] = [float("inf")] * len(adj)  # type: ignore
    prev = [-1] * len(adj)
    visited = [False] * len(adj)
    dist[s] = 0
    pq: PriorityQueue = PriorityQueue()
    pq.put((0, s))

    while not pq.empty():
        u = pq.get()[1]

        if not visited[u]:
            visited[u] = True
            for i in range(len(adj[u])):
                v = adj[u][i]
                if dist[v] > dist[u] + cost[u][i]:
                    dist[v] = dist[u] + cost[u][i]
                    prev[v] = u
                    pq.put((-dist[v], v))

    return dist[t] if dist[t] != float("inf") else -1


if __name__ == "__main__":
    n, m = list(map(int, input().split()))

    edges = [tuple(map(int, input().split())) for _ in range(m)]
    adj: List[List[int]] = [[] for _ in range(n)]
    cost: List[List[int]] = [[] for _ in range(n)]
    for a, b, w in edges:
        adj[a - 1].append(b - 1)
        cost[a - 1].append(w)

    u, v = map(int, input().split())
    print(distance(adj, cost, u - 1, v - 1))

File: graph_algorithms/week4_fastest_paths/1_minimum_flight_cost/dijkstra.cpp
#include <iostream>
#include <queue>
#include <vector>

using std::pair;
using std::priority_queue;
using std::queue;
using std::vector;

long long distance(vector<vector<int> > &adj, vector<vector<int> > &cost, int s, int t) {
  vector<long long> dist(adj.size(), __LONG_LONG_MAX__);
  vector<int> prev(adj.size(), -1);
  vector<bool> visited(adj.size(), false);
  dist[s] = 0;
  priority_queue<pair<long long, int> > pq;
  pq.push(std::make_pair(0, s));
  while (!pq.empty()) {
    int u = pq.top().second;
    pq.pop();

    if (!visited[u]) {
      visited[u] = true;
      for (int i = 0; i < adj[u].size(); i++) {
        int v = adj[u][i];
        if (dist[v] > dist[u] + cost[u][i]) {
          dist[v] = dist[u] + cost[u][i];
          prev[v] = u;
          pq.push(std::make_pair(-dist[v], v));
        }
      }
    }
  }
  return dist[t] == __LONG_LONG_MAX__ ? -1 : dist[t];
}

int main() {
  int n, m;
  std::cin >> n >> m;

  vector<vector<int> > adj(n, vector<int>());
  vector<vector<int> > cost(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y, w;
    std::cin >> x >> y >> w;
    adj[x - 1].push_back(y - 1);
    cost[x - 1].push_back(w);
  }

  int u, v;
  std::cin >> u >> v;
  std::cout << distance(adj, cost, u - 1, v - 1) << std::endl;
}
